mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/generation/PropertyType.java

```diff
@@ -0,0 +1,183 @@
+import com.datastax.oss.driver.api.mapper.annotations.Entity;
+import com.datastax.oss.driver.internal.mapper.processor.ProcessorContext;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeName;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+
+/**
+ * Wraps the declared type of an entity property (or DAO method parameter) that will be injected
+ * into a {@link SettableByName}, or extracted from a {@link GettableByName}.
+ *
+ * <p>The goal is to detect if the type contains other mapped entities, that must be translated into
+ * UDT values.
+ */
+public abstract class PropertyType {
+
+  private static final ClassName UDT_VALUE_CLASS_NAME = ClassName.get(UdtValue.class);
+  public static final ClassName LIST_CLASS_NAME = ClassName.get(List.class);
+  public static final ClassName SET_CLASS_NAME = ClassName.get(Set.class);
+  public static final ClassName MAP_CLASS_NAME = ClassName.get(Map.class);
+
+  public static PropertyType parse(TypeMirror typeMirror, ProcessorContext context) {
+    if (typeMirror.getKind() == TypeKind.DECLARED) {
+      DeclaredType declaredType = (DeclaredType) typeMirror;
+      if (declaredType.asElement().getAnnotation(Entity.class) != null) {
+        return new SingleEntity(declaredType);
+      } else if (context.getClassUtils().isList(declaredType)) {
+        PropertyType elementType = parse(declaredType.getTypeArguments().get(0), context);
+        return (elementType instanceof Simple)
+            ? new Simple(typeMirror)
+            : new EntityList(elementType);
+      } else if (context.getClassUtils().isSet(declaredType)) {
+        PropertyType elementType = parse(declaredType.getTypeArguments().get(0), context);
+        return (elementType instanceof Simple)
+            ? new Simple(typeMirror)
+            : new EntitySet(elementType);
+      } else if (context.getClassUtils().isMap(declaredType)) {
+        PropertyType keyType = parse(declaredType.getTypeArguments().get(0), context);
+        PropertyType valueType = parse(declaredType.getTypeArguments().get(1), context);
+        return (keyType instanceof Simple && valueType instanceof Simple)
+            ? new Simple(typeMirror)
+            : new EntityMap(keyType, valueType);
+      }
+    }
+    return new Simple(typeMirror);
+  }
+
+  public abstract TypeName asTypeName();
+
+  /**
+   * Returns the name of the type we will convert to before saving to the database; that is,
+   * replacing every entity class by {@code UdtValue}.
+   */
+  public abstract TypeName asRawTypeName();
+
+  /**
+   * A type that does not contain any mapped entity.
+   *
+   * <p>Note that it can still be a collection, for example {@code Map<String, List<Integer>>}.
+   */
+  public static class Simple extends PropertyType {
```

Comment:
Maybe the `PropertyType` should be enum instead of the class? It could be simpler for a caller because the compiler will check that all enum cases were covered and give warning otherwise (just a suggestion)

Reply:
`PropertyType` does not have a predefined set of values, because it wraps an underlying type mirror.
For example, a `PropertyType.Simple` can correspond to an `int`, a `String`, a `Map<String, List<Integer>>`, etc.